File: leetcode/双指针/0086_Partition_List/Solution.py

class Solution(object):
    def partition(self, head, x):
        """
        :type head: ListNode
        :type x: int
        :rtype: ListNode
        """
        dummy = ListNode(-1, head)
        dummy_tmp = ListNode(-1)
        p1 = dummy
        p2 = dummy_tmp

        while head:
            if head.val < x:
                p1 = p1.next
                head = head.next
            else:
                p2.next = head
                p2 = p2.next

                head = head.next
                p1.next = head

                p2.next = None
        
        p1.next = dummy_tmp.next

        return dummy.next


class ListNode(object):
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

File: leetcode/双指针/0086_Partition_List/Solution.java
public class Solution {

    public ListNode partition(ListNode head, int x) {
        ListNode dummy = new ListNode(-1, head);  // 别忘了 next 指向 head
        ListNode dummy_tmp = new ListNode(-1);
        ListNode pre = dummy, pre_tmp = dummy_tmp;

        while (head != null) {
            if (head.val < x) {
                pre = pre.next;
                head = head.next;
            } else {
                pre_tmp.next = head;
                pre_tmp = pre_tmp.next;

                head = head.next;
                pre.next = head;
                // 注意: 这里不能移动 pre = pre.next, 否则 pre 就指向 head 了
                
                pre_tmp.next = null; // 必须要断开新链
            }
        }

        // 将 两条链 连接上
        pre.next = dummy_tmp.next;

        return dummy.next;
    }

    static class ListNode {
        int val;
        ListNode next;
        ListNode() {}
        ListNode(int val) { this.val = val; }
        ListNode(int val, ListNode next) { this.val = val; this.next = next; }
    }
}
